modules/page-memory/src/test/java/org/apache/ignite/internal/pagememory/persistence/checkpoint/CheckpointReadWriteLockTest.java

```diff
@@ -0,0 +1,206 @@
+                fail(e);
+            }
+
+            assertEquals(0, lock0.getReadHoldCount());
+            assertEquals(0, lock1.getReadHoldCount());
+            assertEquals(0, lock2.getReadHoldCount());
+        }).get(1, TimeUnit.SECONDS);
+
+        lock2.writeUnlock();
+
+        runAsync(() -> {
+            try {
+                assertTrue(lock0.tryReadLock());
+                assertTrue(lock1.tryReadLock(1, TimeUnit.MILLISECONDS));
+
+                assertTrue(lock2.tryReadLock());
+                assertTrue(lock2.tryReadLock(1, TimeUnit.MILLISECONDS));
+
+                assertEquals(1, lock0.getReadHoldCount());
+                assertEquals(1, lock1.getReadHoldCount());
+                assertEquals(2, lock2.getReadHoldCount());
+
+                lock0.readUnlock();
+                lock1.readUnlock();
+                lock2.readUnlock();
+
+                assertEquals(0, lock0.getReadHoldCount());
+                assertEquals(0, lock1.getReadHoldCount());
+                assertEquals(1, lock2.getReadHoldCount());
+
+                lock2.readUnlock();
+
+                assertEquals(0, lock2.getReadHoldCount());
+            } catch (InterruptedException e) {
+                fail(e);
+            }
+        }).get(1, TimeUnit.SECONDS);
+
+        lock2.writeLock();
+
+        assertEquals(1, lock0.getReadHoldCount());
+        assertEquals(1, lock1.getReadHoldCount());
+        assertEquals(0, lock2.getReadHoldCount());
+
+        lock0.readUnlock();
+        lock1.readUnlock();
+        lock2.readUnlock();
+
+        assertEquals(0, lock0.getReadHoldCount());
+        assertEquals(0, lock1.getReadHoldCount());
+        assertEquals(0, lock2.getReadHoldCount());
+    }
+
+    @Test
+    void testCheckpointLockIsHeldByThread() throws Exception {
+        CheckpointReadWriteLock lock0 = new CheckpointReadWriteLock(new ReentrantReadWriteLockWithTracking());
+        CheckpointReadWriteLock lock1 = new CheckpointReadWriteLock(new ReentrantReadWriteLockWithTracking());
+        CheckpointReadWriteLock lock2 = new CheckpointReadWriteLock(new ReentrantReadWriteLockWithTracking());
+
+        assertFalse(lock0.checkpointLockIsHeldByThread());
+        assertFalse(lock1.checkpointLockIsHeldByThread());
+        assertFalse(lock2.checkpointLockIsHeldByThread());
+
+        lock0.writeLock();
+        lock1.readLock();
+
+        assertTrue(lock0.checkpointLockIsHeldByThread());
+        assertTrue(lock1.checkpointLockIsHeldByThread());
+
+        runAsync(
+                () -> assertTrue(lock2.checkpointLockIsHeldByThread()),
+                newSingleThreadExecutor(new NamedThreadFactory("checkpoint-runner"))
+        ).get(1, TimeUnit.SECONDS);
+
+        runAsync(() -> {
+            assertFalse(lock0.checkpointLockIsHeldByThread());
+            assertFalse(lock1.checkpointLockIsHeldByThread());
+            assertFalse(lock2.checkpointLockIsHeldByThread());
+        }).get(1, TimeUnit.SECONDS);
+
+        runAsync(() -> {
```

Comment:
Why not "IgniteTestUtils#runAsync(...)"? Did you have some specific reasons?

Reply:
Thread name, for example. And, all threads started from tests should be stopped somewhere in "after test". It's not implemented yet, but that would be much easier with our own factories.
We can ignore it for now, I guess, and refactor everything in the future. Just keep in mind, there always may be reasons to use custom methods.

Reply:
Change it.